pkg/controller/externalnode/controller.go

```diff
@@ -0,0 +1,386 @@
+		return
+	}
+	if err := c.reconcileExternalNode(); err != nil {
+		klog.ErrorS(err, "Failed to reconcile ExternalNodes")
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// reconcileExternalNode reconciles all the existing ExternalNodes and cleans up the stale ExternalEntities.
+func (c *ExternalNodeController) reconcileExternalNode() error {
+	externalNodes, err := c.externalNodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	enUIDEENameMap := make(map[types.UID]string)
+	for _, en := range externalNodes {
+		if err = c.addExternalNode(en); err != nil {
+			return err
+		}
+		eeName := genExternalEntityName(en)
+		enUIDEENameMap[en.UID] = eeName
+	}
+	externalEntities, err := c.externalEntityLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, ee := range externalEntities {
+		if (len(ee.OwnerReferences) > 0) && (ee.OwnerReferences[0].Kind == "ExternalNode") {
+			// Clean up stale ExternalEntities when ExternalNode no longer exists or
+			// when interface[0] name is changed.
+			if eeName, ok := enUIDEENameMap[ee.OwnerReferences[0].UID]; !ok || (ok && (eeName != ee.Name)) {
+				err = c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Delete(context.TODO(), ee.Name, metav1.DeleteOptions{})
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the work queue.
+func (c *ExternalNodeController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *ExternalNodeController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in ExternalNode work queue but got %#v", obj)
+		return true
+	} else if err := c.syncExternalNode(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing ExternalNode", "ExternalNode", key)
+	}
+	return true
+}
+
+func (c *ExternalNodeController) syncExternalNode(key string) error {
+	namespace, name, err := splitKeyFunc(key)
+	if err != nil {
+		// This err should not occur.
+		return err
+	}
+	en, err := c.externalNodeLister.ExternalNodes(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return c.deleteExternalNode(namespace, name)
+	}
+
+	preEn, exists, _ := c.syncedExternalNode.GetByKey(key)
+	if !exists {
+		return c.addExternalNode(en)
+	} else {
+		return c.updateExternalNode(preEn.(*v1alpha1.ExternalNode), en)
+	}
+}
+
+func (c *ExternalNodeController) addExternalNode(en *v1alpha1.ExternalNode) error {
```

Comment:
Better to add some comments about what this func does.

Reply:
Thanks, I have added some comments here.